### ISSUE TITLE
VRRP master preeemt not working correctly on startup when vrrp_startup_delay is used.

### ISSUE THREAD

**Issue by cvachoucek** [acassen/keepalived#2098]
Reducing the master-down-timer (vrrp->ms_down_timer) value, while starting in BACKUP state and incoming advert is ignored, is wrong. It does not take into account running vrrp_startup_delay, so the vrrp->ms_down_timer is in fact increased, not decreased (until the vrrp_startup_delay does not expire). This results in delaying the preemption for about vrrp_startup_delay with every incoming ignored advert packet, and log full of:

Wed Feb  9 07:30:58 2022: (VI_1) received lower priority (70) advert from 10.0.0.105 - discarding

Attaching patch - the vrrp->ms_down_timer should not be re-calculated while the vrrp_startup_delay is running.

[vrrp-startup-delay-fix.txt](https://github.com/acassen/keepalived/files/8030686/vrrp-startup-delay-fix.txt)

**Comment by cvachoucek** [comment]
Yes I'd say completely ignoring incoming adverts while vrrp_startup_delay is running is a better approach.
For the patch - I get a compile error when the patch is applied (on the master sources):

```
vrrp_daemon.c: In function ‘start_vrrp’:
vrrp_daemon.c:601:21: error: ‘thread’ undeclared (first use in this function); did you mean ‘_thread’?
  601 |    thread_add_timer(thread->master, delayed_start_clear_thread, NULL, global_data->vrrp_startup_delay);
      |                     ^~~~~~
      |                     _thread
vrrp_daemon.c:601:21: note: each undeclared identifier is reported only once for each function it appears in
```

I've replaced the thread->master by just master to make it compile.

Then the fd variable in vrrp_read_dispatcher_thread() function needs to be initialized, I've initialized it to -1.
After these changes, the code compiles fine, but it does not work, the log looks like this:

```
Thu Feb 10 08:52:34 2022: Starting Keepalived v2.2.7 (02/08,2022), git commit v2.2.7-10-g1f375e83+
Thu Feb 10 08:52:34 2022: Command line: './sbin/keepalived' '-nDlG' '-f' './keepalived.conf'
Thu Feb 10 08:52:34 2022: Opening file './keepalived.conf'.
Thu Feb 10 08:52:34 2022: Configuration file ./keepalived.conf
Thu Feb 10 08:52:34 2022: Starting VRRP child process, pid=2503900
Thu Feb 10 08:52:34 2022: Registering Kernel netlink reflector
Thu Feb 10 08:52:34 2022: Registering Kernel netlink command channel
Thu Feb 10 08:52:34 2022: Delaying startup for 3 seconds
Thu Feb 10 08:52:34 2022: Assigned address 10.0.0.10 for interface enp0s31f6
Thu Feb 10 08:52:34 2022: Assigned address fe80::c015:eff0:6b29:bcc5 for interface enp0s31f6
Thu Feb 10 08:52:34 2022: Registering gratuitous ARP shared channel
Thu Feb 10 08:52:34 2022: (VI_1) removing VIPs.
Thu Feb 10 08:52:34 2022: Startup complete
Thu Feb 10 08:52:34 2022: (VI_1) Entering BACKUP STATE (init)
Thu Feb 10 08:52:34 2022: VRRP sockpool: [ifindex(  2), family(IPv4), proto(112), fd(11,12) multicast, address(224.0.0.18)]
Thu Feb 10 08:52:34 2022: scheduler: No read thread bound on fd:11 (fl:0x0001)
Thu Feb 10 08:52:34 2022: scheduler: No read thread bound on fd:11 (fl:0x0001)
Thu Feb 10 08:52:34 2022: scheduler: No read thread bound on fd:11 (fl:0x0001)
Thu Feb 10 08:52:34 2022: scheduler: No read thread bound on fd:11 (fl:0x0001)
... and so on (endless)
```

The incoming packets should be read and discarded IMHO.

**Comment by pqarmitage** [comment]
Apologies for the errors in the patch. Yes, replacing `thread->master` with `master` is correct. However, rather that adding `fd = -1` it should have been `fd = thread->u.f.fd`, but this still has a problem since any received adverts are not read from the receive queue, and so when the delayed start time expires, keepalived process all the adverts that had been ignored during the delayed start.

Attached is a revised patch. It now reads the adverts that are received during the delayed start, checks and reports any receive errors, and then discards the adverts. There is no need to protect the read_timeout because the thread timeout is set to expire after the delayed start.

Here is that patch - I have actually tested it this time!

[vrrp-startup-delay-fix-v3.txt](https://github.com/acassen/keepalived/files/8039408/vrrp-startup-delay-fix-v3.txt)

**Comment by cvachoucek** [comment]
Now it seems to work fine, as expected.

```
Thu Feb 10 09:49:31 2022: Starting Keepalived v2.2.7 (02/08,2022), git commit v2.2.7-10-g1f375e83+
Thu Feb 10 09:49:31 2022: Command line: './sbin/keepalived' '-nDlG' '-f' './keepalived.conf'
Thu Feb 10 09:49:31 2022: Opening file './keepalived.conf'.
Thu Feb 10 09:49:31 2022: Configuration file ./keepalived.conf
Thu Feb 10 09:49:31 2022: Starting VRRP child process, pid=2509417
Thu Feb 10 09:49:31 2022: Registering Kernel netlink reflector
Thu Feb 10 09:49:31 2022: Registering Kernel netlink command channel
Thu Feb 10 09:49:31 2022: Delaying startup for 3 seconds
Thu Feb 10 09:49:31 2022: Assigned address 10.0.0.10 for interface enp0s31f6
Thu Feb 10 09:49:31 2022: Assigned address fe80::c015:eff0:6b29:bcc5 for interface enp0s31f6
Thu Feb 10 09:49:31 2022: Registering gratuitous ARP shared channel
Thu Feb 10 09:49:31 2022: (VI_1) removing VIPs.
Thu Feb 10 09:49:31 2022: Startup complete
Thu Feb 10 09:49:31 2022: (VI_1) Entering BACKUP STATE (init)
Thu Feb 10 09:49:31 2022: VRRP sockpool: [ifindex(  2), family(IPv4), proto(112), fd(11,12) multicast, address(224.0.0.18)]
Thu Feb 10 09:49:34 2022: (VI_1) received lower priority (90) advert from 10.0.0.100 - discarding
Thu Feb 10 09:49:34 2022: (VI_1) received lower priority (90) advert from 10.0.0.100 - discarding
Thu Feb 10 09:49:34 2022: (VI_1) received lower priority (90) advert from 10.0.0.100 - discarding
Thu Feb 10 09:49:34 2022: (VI_1) received lower priority (90) advert from 10.0.0.100 - discarding
Thu Feb 10 09:49:34 2022: (VI_1) Receive advertisement timeout
Thu Feb 10 09:49:34 2022: (VI_1) Entering MASTER STATE
Thu Feb 10 09:49:34 2022: (VI_1) setting VIPs.
Thu Feb 10 09:49:34 2022: (VI_1) Sending/queueing gratuitous ARPs on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:34 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:34 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:34 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:34 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:34 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:39 2022: (VI_1) Sending/queueing gratuitous ARPs on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:39 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:39 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:39 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:39 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
Thu Feb 10 09:49:39 2022: Sending gratuitous ARP on enp0s31f6 for 10.0.0.99
```

Thank you.

**Comment by pqarmitage** [comment]
Commit 845181c applies the latest patch above. I have also added logging when the delayed start completes.
